chasm/lib/scheduler/invoker_tasks.go

```diff
@@ -0,0 +1,610 @@
+	scheduler1 "go.temporal.io/server/service/worker/scheduler"
+	"go.uber.org/fx"
+	"google.golang.org/protobuf/types/known/timestamppb"
+)
+
+type (
+	InvokerTaskExecutorOptions struct {
+		fx.In
+
+		Config         *Config
+		MetricsHandler metrics.Handler
+		BaseLogger     log.Logger
+		HistoryClient  resource.HistoryClient
+		FrontendClient workflowservice.WorkflowServiceClient
+	}
+
+	InvokerExecuteTaskExecutor struct {
+		InvokerTaskExecutorOptions
+	}
+
+	InvokerProcessBufferTaskExecutor struct {
+		InvokerTaskExecutorOptions
+	}
+
+	// Per-task context.
+	invokerTaskExecutorContext struct {
+		context.Context
+
+		actionsTaken int
+		maxActions   int
+	}
+
+	rateLimitedError struct {
+		// The requested interval to delay processing by rescheduilng.
+		delay time.Duration
+	}
+)
+
+const (
+	// Lower bound for the deadline in which buffered actions are dropped.
+	startWorkflowMinDeadline = 5 * time.Second
+
+	// Because the catchup window doesn't apply to a manual start, pick a custom
+	// execution deadline before timing out a start.
+	manualStartExecutionDeadline = 1 * time.Hour
+
+	// Upper bound on how many times starting an individual buffered action should be retried.
+	InvokerMaxStartAttempts = 10 // TODO - dial this up/remove it
+)
+
+var (
+	errRetryLimitExceeded       = errors.New("retry limit exceeded")
+	_                     error = &rateLimitedError{}
+)
+
+func NewInvokerExecuteTaskExecutor(opts InvokerTaskExecutorOptions) *InvokerExecuteTaskExecutor {
+	return &InvokerExecuteTaskExecutor{
+		InvokerTaskExecutorOptions: opts,
+	}
+}
+
+func NewInvokerProcessBufferTaskExecutor(opts InvokerTaskExecutorOptions) *InvokerProcessBufferTaskExecutor {
+	return &InvokerProcessBufferTaskExecutor{
+		InvokerTaskExecutorOptions: opts,
+	}
+}
+
+func (e *InvokerExecuteTaskExecutor) Validate(
+	_ chasm.Context,
+	invoker *Invoker,
+	_ chasm.TaskAttributes,
+	_ *schedulespb.InvokerExecuteTask,
+) (bool, error) {
+	// If another execute task already happened to kick everything off, we don't need
+	// this one.
+	eligibleStarts := invoker.getEligibleBufferedStarts()
+	valid := len(invoker.GetTerminateWorkflows())+
+		len(invoker.GetCancelWorkflows())+
+		len(eligibleStarts) > 0
+	return valid, nil
+}
+
+func (e *InvokerExecuteTaskExecutor) Execute(
+	ctx context.Context,
+	invokerRef chasm.ComponentRef,
+	_ chasm.TaskAttributes,
+	_ *schedulespb.InvokerExecuteTask,
+) error {
+	var result executeResult
+
+	invoker, err := chasm.ReadComponent(
+		ctx,
+		invokerRef,
+		func(i *Invoker, _ chasm.Context, _ struct{}) (*Invoker, error) {
+			return i, nil
+		},
+		struct{}{},
+	)
+	if err != nil {
+		return fmt.Errorf("%w: %w",
+			serviceerror.NewInternal("Failed to read component"),
+			err)
+	}
+
+	schedulerRef := chasm.ComponentRef{
+		EntityKey: invokerRef.EntityKey,
+	}
+	scheduler, err := chasm.ReadComponent(
+		ctx,
+		schedulerRef,
+		func(s *Scheduler, _ chasm.Context, _ struct{}) (*Scheduler, error) {
+			return s, nil
```

Comment:
shall we have a deep copy here? same Q for the ReadComponent logic for invoker above.

Reply:
Also, can/should we return both scheduler and invoker in one transition?